### DISCUSSION TITLE
Support for direct access to S3 resources

### DISCUSSION THREAD

**Discussion by sahuguet** [SPARQL-Anything/sparql.anything#438] [Ideas]
Assuming this is not implemented yet (I could not find it in the doc), I would like to reference s3 resources directly without having to download them locally.

A good example is https://duckdb.org/docs/extensions/httpfs.html where S3 resources can be referenced seamlessly from the query.

**Comment by luigi-asprino** [comment]
Hi,

You're right. It's not implemented yet, but it would be very nice to have it!
I'm not familiar with S3 resources, but I assume there should be a way to open an input stream from them.
There seems to be a Java library for this https://github.com/aws/aws-sdk-java/tree/master

**Reply by sahuguet** [reply]
The kind of logic to implement to access S3 over http can be found at https://github.com/duckdb/duckdb/blob/main/extension/httpfs/s3fs.cpp#L19 . But it is probably cleaner and simpler to use a Java library that does the work.

**Reply by kvistgaard** [reply]
Hi @luigi-asprino , what are the intentions regarding this?
The lack of this feature is currently blocking the use of SA in one of my projects.

**Reply by luigi-asprino** [reply]
Hi, 

I wasn't able to look into it. I'm not familiar with the type of resource, so if you (or anyone else) can help me in understanding how to read such a resource from Java, I would be happy to integrate it into SA.

**Reply by MathiasVDA** [reply]
I have experience with handling (CRUD) S3 objects in .net, but I'm not so versed in java. I did a little bit of lookup and maybe this can help:
- You already found the library [AWS SDK for Java](https://aws.amazon.com/sdk-for-java/) to interact with S3. That should allow you to open an object in S3 using the S3 client’s getObject method, which returns an S3ObjectInputStream.
- You will need to handle authentication (in the service call of SA?)

I've done some q&a with copilot in this project and this might be useful:
- To keep things modular, you could create a resource handler or input stream provider class for S3 URIs, similar to how other resource types are managed in SPARQL Anything.
- The implementation would involve recognizing s3:// URIs and delegating to the AWS SDK to fetch the stream.

**Reply by luigi-asprino** [reply]
thanks, is there any way to test it? any dummy resource that I can retrieve from S3?

**Reply by MathiasVDA** [reply]
There are some options to run a local S3 storage solution like min.io, easy to do with docker: https://hub.docker.com/r/minio/minio

Let me see if I can spin one up on the web for you to test

**Reply by MathiasVDA** [reply]
I've set up a minio instance on one of my webservers for you to experiment. I'll send you the login details over PM @luigi-asprino or anyone else interested in contributing. 

I've added a bucket "sparql-anything" and a file "some.csv" to it. 
S3 storage endpoint is https://minio.matdata.eu and the webGUI (console) is at https://minio-console.matdata.eu

**Reply by luigi-asprino** [reply]
great, my personal email is luigi.asprino@gmail.com

**Reply by MathiasVDA** [reply]
yes I know, you already have mail :)

**Reply by luigi-asprino** [reply]
I transform this discussion into a issue to keep track of the development.

**Reply by MathiasVDA** [reply]
Here's a small python snippet that works on my computer after `pip install pandas boto` :


```
"""
Simple example for accessing MinIO S3 storage and reading CSV
"""

import boto3
import pandas as pd
import io

# Please fill in your MinIO credentials
MINIO_USERNAME = ""  # Your MinIO access key ID
MINIO_PASSWORD = ""  # Your MinIO secret access key

def simple_s3_csv_reader():
    """Simple function to read CSV from MinIO S3"""
    
    # Create S3 client
    s3 = boto3.client(
        's3',
        endpoint_url='https://minio.matdata.eu',
        aws_access_key_id=MINIO_USERNAME,
        aws_secret_access_key=MINIO_PASSWORD
    )
    
    try:
        # Download the file
        response = s3.get_object(Bucket='sparql-anything', Key='some.csv')
        
        # Read CSV into pandas DataFrame
        df = pd.read_csv(io.BytesIO(response['Body'].read()))
        
        print("CSV loaded successfully!")
        print(f"Shape: {df.shape}")
        print("\nFirst 5 rows:")
        print(df.head())
        
        return df
        
    except Exception as e:
        print(f"Error: {e}")
        return None

if __name__ == "__main__":
    if not MINIO_USERNAME or not MINIO_PASSWORD:
        print("Please fill in your MinIO credentials first!")
    else:
        df = simple_s3_csv_reader()
```

**Reply by enridaga** [reply]
> I transform this discussion into a issue to keep track of the development.

This feature was added in https://github.com/SPARQL-Anything/sparql.anything/issues/560
